responses/__init__.py

```diff
@@ -241,6 +246,14 @@ def __iter__(self) -> Iterator[Call]:
     def __len__(self) -> int:
         return len(self._calls)
 
+    @overload
+    def __getitem__(self, idx: int) -> Call:
```

Comment:
Recent version of mypy doesn't require overload, why do we necessarily need it?

Reply:
I have `mypy==1.6.1`, which is the most current version and it still throws:

```
error: Item "list[Call]" of "Call | list[Call]" has no attribute "request"  [union-attr]
```

Changes in this PR should fix that. I can't really imagine how would `mypy` guess the appropriate combination of input and output types without overload.

Reply:
The previous CI passed just fine with mypy, although I saw some flakiness in results in the last couple of months

What is the difference then?

Reply:
That will be because `mypy` isn't being run against the tests. The definition of `CallList`'s `__getitem__` method is fine. It's where it's used that becomes a problem. You can reproduce this locally removing the line that excludes the tests from `mypy` in the .ini file.

This is why I've suggested running static analysis on the tests.

Reply:
then we need to indeed enabling mypy on tests to validate this PR and all the future PRs

Reply:
Is it worth getting this out to fix the regression, and then dealing with enabling static analysis in the tests in a separate PR? My original intention was to do that here but after removing the exclusion from the mypy config file there were *754* errors. Fixing those is going to make the diff really noisy and feels like a separate issue (IMO).